onnxruntime/contrib_ops/cpu/quantization/moe_quantization_cpu.cc

```diff
@@ -0,0 +1,438 @@
+          float* dequant_fc1_expert = dequant_fc1_weights.get() + expert_idx * moe_params.hidden_size * (is_4bit ? fc1_output_size : moe_params.inter_size * act_multiplier);
+
+          const int64_t output_cols = is_4bit ? fc1_output_size : moe_params.inter_size * act_multiplier;
+          for (int64_t out_col = 0; out_col < output_cols; ++out_col) {
+            for (int64_t in_col = 0; in_col < moe_params.hidden_size; ++in_col) {
+              size_t linear_idx = static_cast<size_t>(out_col * moe_params.hidden_size + in_col);
+              dequant_fc1_expert[linear_idx] = DequantizeWeight(fc1_expert_weights, linear_idx, linear_idx, fc1_expert_scales, out_col);
+            }
+          }
+        }
+      });
+
+  // Dequantize FC2 weights for all experts
+  concurrency::ThreadPool::TryParallelFor(
+      thread_pool, static_cast<std::ptrdiff_t>(moe_params.num_experts),
+      static_cast<double>(std::max<int64_t>(1, moe_params.num_experts / num_threads)),
+      [&](ptrdiff_t expert_start, ptrdiff_t expert_end) {
+        for (int64_t expert_idx = expert_start; expert_idx < expert_end; ++expert_idx) {
+          const uint8_t* fc2_expert_weights = fc2_weights_data + expert_idx * fc2_weight_stride;
+          const float* fc2_expert_scales = fc2_scales_data + expert_idx * moe_params.hidden_size;
+          float* dequant_fc2_expert = dequant_fc2_weights.get() + expert_idx * moe_params.inter_size * moe_params.hidden_size;
+
+          for (int64_t out_col = 0; out_col < moe_params.hidden_size; ++out_col) {
+            for (int64_t in_col = 0; in_col < moe_params.inter_size; ++in_col) {
+              size_t linear_idx = static_cast<size_t>(out_col * moe_params.inter_size + in_col);
+              dequant_fc2_expert[linear_idx] = DequantizeWeight(fc2_expert_weights, linear_idx, linear_idx, fc2_expert_scales, out_col);
+            }
+          }
+        }
+      });
+
+  // Process tokens in parallel
+  concurrency::ThreadPool::TryParallelFor(
+      thread_pool, static_cast<std::ptrdiff_t>(moe_params.num_rows),
+      static_cast<double>(std::max<int64_t>(1, moe_params.num_rows / num_threads)),
+      [&](ptrdiff_t start_token, ptrdiff_t end_token) {
+        const int64_t thread_id = start_token / ((moe_params.num_rows + num_threads - 1) / num_threads);
+        const int64_t thread_fc1_size = is_4bit ? (moe_params.inter_size * (is_swiglu ? 2 : 1)) : (moe_params.inter_size * act_multiplier);
+        float* thread_fc1_output = thread_fc1_buffers.get() + thread_id * thread_fc1_size;
+        float* thread_fc2_output = thread_fc2_buffers.get() + thread_id * moe_params.hidden_size;
+        float* thread_local_results = thread_results.get() + thread_id * moe_params.num_rows * moe_params.hidden_size;
+        float* thread_bias_buffer = thread_bias_buffers.get() + thread_id * max_bias_size;
+
+        // Process each token in this thread's range
+        for (int64_t token_idx = start_token; token_idx < end_token; ++token_idx) {
```

Comment:
Keeping the type as `std::ptrdiff_t` is safer.

```suggestion
        for (std::ptrdiff_t token_idx = start_token; token_idx < end_token; ++token_idx) {
```

We can then use [`SafeInt`](https://learn.microsoft.com/en-us/cpp/safeint/safeint-library?view=msvc-170) to prevent integer overflow when using `expert_idx`. Here is an example.

https://github.com/microsoft/onnxruntime/blob/4a8a289e635f5beabd33d111fd7d31f549300682/onnxruntime/contrib_ops/cpu/bert/attention_cpu_base.h#L260